dig_test.go

```diff
@@ -2455,214 +2456,266 @@ func TestFailingFunctionDoesNotCreateInvalidState(t *testing.T) {
 	}), "second invoke must fail")
 }
 
-func TestDotGraph(t *testing.T) {
-	type T1 struct{}
-	type T2 struct{}
-	type T3 struct{}
-	type T4 struct{}
+func compareCtor(t *testing.T, expected *dot.Ctor, ctor *dot.Ctor) {
+	assert.Equal(t, expected.Name, ctor.Name)
+	assert.Equal(t, expected.Package, ctor.Package)
+	assert.Equal(t, path.Base(expected.File), path.Base(ctor.File))
+	assert.Equal(t, expected.Params, ctor.Params)
+	assert.Equal(t, expected.Results, ctor.Results)
+	assert.NotZero(t, ctor.Line)
+}
 
-	e1, e2, e3, e4 := &dot.Node{Type: "dig.T1"}, &dot.Node{Type: "dig.T2"}, &dot.Node{Type: "dig.T3"}, &dot.Node{Type: "dig.T4"}
+func compareCtors(t *testing.T, expected []*dot.Ctor, ctors []*dot.Ctor) {
+	for i := range ctors {
+		compareCtor(t, expected[i], ctors[i])
+	}
+}
 
-	t.Parallel()
+func TestDotGraph(t *testing.T) {
+	type t1 struct{}
+	type t2 struct{}
+	type t3 struct{}
+	type t4 struct{}
 
-	t.Run("single edge", func(t *testing.T) {
-		expected := []*dot.Edge{
-			{Param: e1, Result: e2},
-		}
+	pkg := "go.uber.org/dig"
+	file := "dig_test.go"
 
-		c := New()
+	e1, e2, e3, e4 := &dot.Node{Type: "dig.t1"}, &dot.Node{Type: "dig.t2"}, &dot.Node{Type: "dig.t3"}, &dot.Node{Type: "dig.t4"}
 
-		c.Provide(func(t1 T1) T2 { return T2{} })
-		assert.Equal(t, expected, c.dg.Edges)
-	})
+	t.Parallel()
 
-	t.Run("more edges", func(t *testing.T) {
-		expected := []*dot.Edge{
-			{Param: e1, Result: e2},
-			{Param: e1, Result: e3},
-			{Param: e2, Result: e4},
-		}
+	t.Run("constructor with one param and one result", func(t *testing.T) {
+		expected := []*dot.Ctor{{
+			Name:    "TestDotGraph.func1.1",
+			Package: pkg,
+			File:    file,
+			Params:  []*dot.Node{e1},
+			Results: []*dot.Node{e2},
+		}}
 
 		c := New()
+		c.Provide(func(A t1) t2 { return t2{} })
+		compareCtors(t, expected, c.dg.Ctors)
+	})
+
+	t.Run("more constructors", func(t *testing.T) {
+		expected := []*dot.Ctor{{
+			Name:    "TestDotGraph.func2.1",
+			Package: pkg,
```

Comment:
formatting nit: for complex literals like this, we prefer the `{` to be on its
own line and the `}` to be with the `,`.

```
[]*dot.Ctor{
    {
        Package: pkg,
        Params: []*dot.Node{e1},
        Results: []*dot.Node{e2},
    },
    {
        // ..
    },
}
```